src/core/config/npo/tracker.rs

```diff
@@ -91,80 +91,117 @@ impl<'a> FieldName<'a> {
-    cache: HashMap<(TypeName<'a>, bool), Chunk<Chunk<FieldName<'a>>>>,
+    // Caches resolved chunks for the specific type
+    // with is_list info since the result is different depending on this flag
+    cache: HashMap<(TypeName<'a>, bool), Chunk<Chunk<Name<'a>>>>,
 }
 
 impl<'a> PathTracker<'a> {
     pub fn new(config: &'a Config) -> PathTracker<'a> {
         PathTracker { config, cache: Default::default() }
     }
 
-    #[allow(clippy::too_many_arguments)]
     fn iter(
         &mut self,
-        path: Chunk<FieldName<'a>>,
+        path: Chunk<Name<'a>>,
         type_name: TypeName<'a>,
         is_list: bool,
-        visited: HashSet<(TypeName<'a>, FieldName<'a>)>,
-    ) -> Chunk<Chunk<FieldName<'a>>> {
-        if let Some(chunks) = self.cache.get(&(type_name, is_list)) {
-            return chunks.clone();
-        }
+    ) -> Chunk<Chunk<Name<'a>>> {
+        let chunks = if let Some(chunks) = self.cache.get(&(type_name, is_list)) {
+            chunks.clone()
+        } else {
+            // set empty value in the cache to prevent infinity recursion
+            self.cache.insert((type_name, is_list), Chunk::new());
+
+            let mut chunks = Chunk::new();
+            if let Some(type_of) = self.config.find_type(type_name.as_str()) {
+                for (name, field) in type_of.fields.iter() {
+                    let chunk_name = Name::Field(FieldName::new(name));
+                    let path = Chunk::new().append(chunk_name);
 
-        let mut chunks = Chunk::new();
-        if let Some(type_of) = self.config.find_type(type_name.as_str()) {
-            for (name, field) in type_of.fields.iter() {
-                let field_name = FieldName::new(name);
-                let path = path.clone().append(field_name);
-                if !visited.contains(&(type_name, field_name)) {
                     if is_list && field.has_resolver() && !field.has_batched_resolver() {
-                        chunks = chunks.append(path.clone());
+                        chunks = chunks.append(path);
                     } else {
-                        let mut visited = visited.clone();
-                        visited.insert((type_name, field_name));
                         let is_list = is_list | field.type_of.is_list();
                         chunks = chunks.concat(self.iter(
                             path,
                             TypeName::new(field.type_of.name()),
                             is_list,
-                            visited,
                         ))
                     }
                 }
             }
-        }
 
-        self.cache.insert((type_name, is_list), chunks.clone());
-        chunks
+            self.cache.insert((type_name, is_list), chunks.clone());
+
+            chunks
+        };
+
+        // chunks contains only paths from the current type.
+        // Prepend every subpath with parent path
+        chunks.map(&mut |chunk| path.clone().concat(chunk.clone()))
     }
```

Comment:
let me explain how the iter is implemented now. The main goal of this code is to find paths to fields from the Query root that in such paths there is some ancestor field of List type and some of the descendant uses resolver without batching.

That's basically executing a DFS for the type graph while preserving information like path and the condition if some of the ancestors is list during the execution. 

That's was already implemented this way, but the caching wasn't done right:
- the cache was storing the paths starting from the root instead of caching intermediate calculations starting from some inner node. We can cache the calculation but it should be done only on subpath level not the whole path because there could multiple ways to connect to the same type and the whole path from Query root will be different in that case. Previously it was showing duplicates with lost information, now it's fixed, see tests.
- visited was only used to prevent cycles, but the same effect could be achieved by using the cache itself if we set empty value before processing descendants and then updating it after the process. The cycles won't appear in the output.

Reply:
@tusharmath please take a look once again. I have added explanation that could help